### ISSUE TITLE
Question: Disabling custom rules

### ISSUE THREAD

**Issue by bnbalint** [pinterest/ktlint#2168]
Hello!

Do you have any documentation on writing a custom rule to properly disable it using an end of line comment? I'm running into an odd issue where I am only able to disable my rule using an end of line comment in certain situations. Below I will include the rule definition, some passing test cases and some failing test cases. If you can point me in a direction to better write the rule to avoid these issues, that would be fantastic!

Rule Definition:
This rule checks that if a certain annotation is present on a function (@RequestMapping) then one of a set of security annotations is also present on the function. It does a hack approach (new to ktlint!) of collecting all annotations on a function and then checking the set. It also stores off the offset of the @RequestMapping annotation so we can tie the lint violation to the correct line.

```kotlin
class EndpointSecurityAnnotationRule : Rule("segmint-ktlint-rules:endpoint-security-annotation") {
    override fun beforeVisitChildNodes(
        node: ASTNode,
        autoCorrect: Boolean,
        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
    ) {
        // if the node is a function, grab the names of all the annotations
        if (node.elementType == FUN) {
            val annotationNames = mutableListOf<String>()

            // grab the MODIFIER_LIST child node (annotations live here)
            val modifierNode = node.findChildByType(MODIFIER_LIST)

            // store the offset for RequestMapping here
            var requestMappingLine = 0

            if (modifierNode != null) {
                val children = modifierNode.getChildren(null)

                for (child in children) {
                    // grab ANNOTATION_ENTRY nodes
                    if (child.elementType == ANNOTATION_ENTRY) {
                        // grab the name from the nested structure
                        val identifier = child.findChildByType(CONSTRUCTOR_CALLEE)
                            ?.findChildByType(TYPE_REFERENCE)
                            ?.findChildByType(USER_TYPE)
                            ?.findChildByType(REFERENCE_EXPRESSION)
                            ?.findChildByType(IDENTIFIER)

                        val name = identifier?.text

                        if (name != null) {
                            annotationNames.add(name)

                            // save the offset, so we can link the error to the correct line
                            if (name == "RequestMapping"){
                                requestMappingLine = identifier.startOffset
                            }
                        }
                    }
                }
            }

            // this is an endpoint
            if ("RequestMapping" in annotationNames) {
                // endpoints must also have one of these security annotations
                if ("SecureEndpoint" !in annotationNames &&
                    "PreAuth" !in annotationNames &&
                    "RequiresPermission" !in annotationNames &&
                    "PartnerSecured" !in annotationNames) {
                    emit(
                        requestMappingLine,
                        "No security annotation found for endpoint",
                        false,
                    )
                }
            }
        }
    }
}
```

Passing Test Cases:
These test cases pass as expected:
```kotlin
    @Test
    fun endpointSecurityAnnotationTest() {

        val code =
            """
            @Timed
            @RequestMapping(
                value = ["/blah"],
                method = [RequestMethod.POST],
                consumes = [JSON]
            )
            @ApiOperation(value = "Send blah")
            @ApiResponses(
                ApiResponse(code = 200, message = "Blah successfully sent"),
            )
            fun sendFeedback(
                @PathVariable("variable") variable: Long,
                @RequestBody body: Body
            ): ResponseEntity<*> {
            }
            """.trimIndent()

        wrappingRuleAssertThat(code)
            .hasLintViolationWithoutAutoCorrect(2, 2, "No security annotation found for endpoint")
    }

    @Test
    fun endpointSecurityAnnotationTest_noError() {

        val code =
            """
            @Timed
            @RequestMapping(
                value = ["/blah"],
                method = [RequestMethod.POST],
                consumes = [JSON]
            )
            @ApiOperation(value = "Send blah")
            @SecureEndpoint
            @ApiResponses(
                ApiResponse(code = 200, message = "Blah successfully sent"),
            )
            fun sendFeedback(
                @PathVariable("variable") variable: Long,
                @RequestBody body: Body
            ): ResponseEntity<*> {
            }
            """.trimIndent()

        wrappingRuleAssertThat(code)
            .hasNoLintViolations()
    }
```

These test cases do NOT pass because there is something above the @RequestMapping annotation:
```kotlin
    @Test
    fun endpointSecurityAnnotationTest_disable() {

        val code =
            """
            // comment!
            @Timed
            @RequestMapping(value = ["/test"], method = [RequestMethod.GET]) // ktlint-disable segmint-ktlint-rules:endpoint-security-annotation
            fun testFunction(
                @RequestParam(param) param: String?,
                @CookieValue(value = idKey, required = false) id: String?
            ): ResponseEntity<*> {
            }
            """.trimIndent()

        wrappingRuleAssertThat(code)
            .hasNoLintViolations()
    }

    @Test
    fun endpointSecurityAnnotationTest_disableWithCommentAbove() {

        val code =
            """
            // comment above
            @RequestMapping(value = ["/**"], method = [RequestMethod.HEAD], params = ["ok"]) // ktlint-disable segmint-ktlint-rules:endpoint-security-annotation
            fun defaultHead(@RequestParam("ok") ok: String?): ResponseEntity<*> {
                return ResponseEntity<Any>(HttpStatus.OK)
            }
            """.trimIndent()

        wrappingRuleAssertThat(code)
            .hasNoLintViolations()
    }

```

However, this test does pass:
```kotlin
    @Test
    fun endpointSecurityAnnotationTest_disableWithoutCommentAbove() {

        val code =
            """
            @RequestMapping(value = ["/**"], method = [RequestMethod.HEAD], params = ["ok"]) // ktlint-disable segmint-ktlint-rules:endpoint-security-annotation
            fun defaultHead(@RequestParam("ok") ok: String?): ResponseEntity<*> {
                return ResponseEntity<Any>(HttpStatus.OK)
            }
            """.trimIndent()

        wrappingRuleAssertThat(code)
            .hasNoLintViolations()
    }
```

**Comment by paul-dingemans** [comment]
The `// ktlint-disable` directive is going to be removed in the next (`1.0`) version of ktlint. The proper way to disable a rule is using `@Suppress("segmint-ktlint-rules:endpoint-security-annotation")`.

But the problem you are facing should not be resolved with an `@Suppress`. Comments directly above a function are really a part of the function definion. The `PsiViewer` plugin of IntelliJ is very valuable in discovering problems like this. For example for code below:
```
            @RequestMapping(value = ["/**"], method = [RequestMethod.HEAD], params = ["ok"]) // ktlint-disable segmint-ktlint-rules:endpoint-security-annotation
            fun defaultHead(@RequestParam("ok") ok: String?): ResponseEntity<*> {
                return ResponseEntity<Any>(HttpStatus.OK)
            }
```
the PSI looks as follows:
![Screenshot 2023-08-11 at 21 37 39](https://github.com/pinterest/ktlint/assets/5195292/5e35b9ee-8efe-4154-8f35-4734616b07a5)

I don't see immediate where the exact problem is in your code. But somehow, it feels brittle to me. You might want to have a look in ktlint standard rules for different approaches how rules can detect problem. In this case I would advise to look at the annotation rule.
